### DISCUSSION TITLE
Accessing XMP data in MP4 files?

### DISCUSSION THREAD

**Discussion by Numpsy** [Zeugma440/atldotnet#257] [Ideas]
Hi,
I see that the roadmap in the readme file says
```
Support for Broadcast wave metadata : aXML and XMP
```

but are there any thoughts on reading or writing XMP data in MP4 files?

I had a really quick go at reading the data in https://github.com/Numpsy/atldotnet/commit/8887deec037183aac034651e4a63ee512dafa8e4 and it seems to work ok (though that test only works for a single UUID field in the file).
I haven't made any attempt at writing data into mp4 files yet.

**Comment by Zeugma440** [comment]
Hi there,

That part of the roadmap is aimed at WAV files, but I can consider starting with MP4 if you need it, sure.
Could you please e-mail me a sample file with XMP data to play with?

Thanks in advance~

PS : I see you're using a wide array of features. What are you building? 😄

**Reply by Numpsy** [reply]
As it stands, I'm just trying to read and write some custom classification properties that are stored inside XMP fields in MP3 and MP4 files.
It's currently done using the C++ Adobe XMP SDK on Windows, but that's a bit painful for portability to other platforms, and read/write in pure .NET seems a mixed bag (e.g. MetadataExtractor is nice for reading from various formats, but it can't update files at all).

So - ATL seems a nice library, and it could almost do what I needed already, so I thought I'd extended to get that last bit :-)

**Reply by Numpsy** [reply]
This file is the 'empty.m4a' file from the test files in this repository, with an XMP entry added by the Adobe C++ SDK:
[empty.m4a.zip](https://github.com/Zeugma440/atldotnet/files/14387589/empty.m4a.zip)

fwiw I'm reading up on the actual file storage formats for this stuff as I go along, and for MP4 I was looking at section 1.2.7 of this Adobe specification document: 
https://github.com/adobe/XMP-Toolkit-SDK/blob/main/docs/XMPSpecificationPart3.pdf

**Reply by Zeugma440** [reply]
Gotcha! I have the same reference here, but the vocabulary Adobe uses is a bit off from what I understood from the Quicktime/MP4 docs, so I'm more comfortable with a sample 👍 

Stay tuned~

**Reply by Zeugma440** [reply]
Okay, generic support for UUID atoms inside M4A/MP4 files is done!

You'll have to use `Track.AdditionalFields["uuid.KEY"] = "value"`.

Key is the UUID itself, e.g. `BE7ACFCB 97A942E8 9C719994 91E3AFAC` for XMP.

XMP support is on the way~

**Reply by Numpsy** [reply]
Nice, I'll have a look tommorow :-) 

> could you give me the subset of properties that you actually use

The test file I posted contains a ```<tmi:Metadata>``` node, and it's that sort of data I'm reading/writing/

> sometimes persisted as XML attributes

The namespace management is something to consider with those sorts of fields - once the 'tmi' (or similar) namespace is there you can add multiple properties in it, but if it's not present then the ```xmlns:XXXXX=``` attribute needs to be added as well

**Reply by Zeugma440** [reply]
Thanks for the heads up ! I'm _almost_ done struggling with namespaces but I still have to :

- Implement WAV support for XMP
- Write a [code snippet](https://github.com/Zeugma440/atldotnet/wiki/3.-Usage-_-Code-snippets#broadcastWave)
- Decide on whether things are sufficiently user-friendly or if we need an additional helper class for XMP formatting

**Reply by Zeugma440** [reply]
Done ! MP4 and WAV support for XMP are now on the main branch.

[Here](https://github.com/Zeugma440/atldotnet/wiki/3.-Usage-_-Code-snippets#xmp)'s the corresponding code snippet.

As for IXML, you have to write the entire path for each of your data inside `AdditionalFields`. ATL takes care of the XML serialization/deserialization. Let me know if that's fine for you or if you need an additional helper class for that.

**Reply by Numpsy** [reply]
Thanks, I'll have a look, but i'm on holiday for a couple of days now so it'll be after that.

**Reply by Numpsy** [reply]
I had a test with the current source code, and the first attempt at read/write seemed to work :-)

A question about the written XMP data - It looks like, If I have an mp4 file which contains XMP with the packet header/footer (```<?xpacket``` etc) and I update that, the the new file doesn't contain thay any more?

Querying because section ```1.2.7.1``` of the Adobe spec 'XMPSpecificationPart3.pdf' spec says ```The remainder of the box is a typical XMP packet, encoded as UTF-8, including packet wrapper and padding.``` (although it does appear that the Adobe C++ SDK will read the data without the packet wrapper)

**Reply by Zeugma440** [reply]
Indeed, I've had issues writing the `<?xpacket` header and footer using `XmlWriter`. I decided to ignore it as it seems to work anyway.

Is it a problem to you?

**Reply by Numpsy** [reply]
If ATL and the C++ XMP SDK can read it then it's ok for me, I just don't know if there are any other tool sthat might be more strict about it

**Reply by Zeugma440** [reply]
I suggest we keep it that way for now. If anyone finds a case where the xpacket wrapping is mandatory, I'll investigate further.

**Reply by Numpsy** [reply]
A question about the writing format for namespaces (I think this might actually be hitting a bug in the C# version of XmpCore / MetadataExtractor and I'll query over there), but for the record:
If I have an M4A file with the initial XMP header of
```
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 5.6.0">
   <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
      <rdf:Description rdf:about=""
            xmlns:xmp="http://ns.adobe.com/xap/1.0/"
            xmlns:xmpDM="http://ns.adobe.com/xmp/1.0/DynamicMedia/"
            xmlns:tmi="http://titus.com/tmi/1.0/">
```
and I modify the XMP with 5.21, then trhe result is this, with the namespaces grouped into the top node:
```
<x:xmpmeta xmlns:x="adobe:ns:meta/" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    xmlns:xmp="http://ns.adobe.com/xap/1.0/" xmlns:xmpDM="http://ns.adobe.com/xmp/1.0/DynamicMedia/"
    xmlns:tmi="http://titus.com/tmi/1.0/" x:xmptk="XMP Core 5.6.0">
    <rdf:RDF>
        <rdf:Description>
```

If I read the second one with the C++ Adobe XMP SDK then it reads the inner data as expected, but with MetdataExtractor (or the .NET port of XmpCore directly) it throws at https://github.com/drewnoakes/xmp-core-dotnet/blob/0f59b87a8e28fa9648856d30737ab8c38098195f/XmpCore/Impl/ParseRdf.cs#L108 and says the data is invalid.

As it happens, section C.2.2 (pages 37-38) of the Adobe specification at https://github.com/adobe/XMP-Toolkit-SDK/blob/main/docs/XMPSpecificationPart1.pdf states that the ```rdf:RDF``` node can't have attributes other than namespaces and the official C++ parser code at https://github.com/adobe/XMP-Toolkit-SDK/blob/5f6fe44f19842ae43dedc3da3fd5d5dafb29e501/XMPCore/source/ParseRDF.cpp#L743 throws if it does, so I *think* from that that the data written by ATL is ok and it's the validation in xmp-core-dotnet that's wrong (not 100% certain though)

**Reply by Zeugma440** [reply]
First thing first, I'm not a LINQ veteran by any means, but I have trouble understanding how
```c
if ( ! xmlNode.attrs.empty() )
```
could be ported to 
```cs
if (!rdfRdfNode.Attributes().Any())
```
so yeah, I also have doubts concerning the implementation (that also seems to violate the specs, btw).

I get that, from `System.xml`'s point of view, a namespace is an `Attribute`, which prevents from writing exactly the same code as in C++. 
One way of fixing that could be something like (not sure that compiles, but you get the idea)
```cs
if (rdfRdfNode.Attributes().Any(!a.Name.StartsWith("xmlns:"))
```
How about that?

If the maintainer isn't happy with that, we could always force ATL to write the `rdf` namespace at the level of the `rdf:RDF` node to make it work. That won't be a problem on my side.

Please let me know~

**Reply by Numpsy** [reply]
According to https://github.com/drewnoakes/metadata-extractor-dotnet/issues/418#issuecomment-2026175946 the C# code was derived from the Java version (which has since been abandoned by Adobe), so it sounds like it's just tripped a rather long standing bug inherited from that, and hopefully we can get that fixed.

From the spec document, I wondered if something like
```csharp
if (rdfRdfNode.Attributes().Any(attr => attr.IsNamespaceDeclaration == false))
```
would be more like it

**Reply by Numpsy** [reply]
>  we could always force ATL to write the rdf namespace at the level of the rdf:RDF node

Hi,
sorry to come back to this after so long, but I have another related issue (and another 'not sure if this is a bug or not') - 

There is a Kotlin port of the XMP SDK at https://github.com/Ashampoo/xmpcore

As it was derived from the same Java source as the C# version, I thought i'd give it a test to see if it has inherited the same issue. Looking at the source it looked like it would, but then it turns out it doesn't get that far - it has some special code trying to work around issues with corrupt XMP data which strips the xmpmeta tags and then tries to parse just the inner rdf nodes, and that fails with the XMP generated by ATL.

Basically, it looks like if you ask it to parse
```
<x:xmpmeta xmlns:x="adobe:ns:meta/" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    xmlns:xmp="http://ns.adobe.com/xap/1.0/" xmlns:xmpDM="http://ns.adobe.com/xmp/1.0/DynamicMedia/"
    xmlns:tmi="http://titus.com/tmi/1.0/" x:xmptk="XMP Core 5.6.0">
    <rdf:RDF>
        <rdf:Description>
```
then it converts that into 
```
    <rdf:RDF>
        <rdf:Description>
```
and then falls over with an ```undefined prefix: rdf``` error.
I don't know if this is a bug, or a different interpretation of the spec, or something else, but it does seem that putting some of the namespace declarations on the rdf node rather than the xmpmeta node would help?

Thanks

**Reply by Zeugma440** [reply]
I'm not good enough at XML to say for sure what can be tolerated and where heresy starts.

Not trying to sneak past that one, but did you ask that question to the library's maintainer?

**Reply by Numpsy** [reply]
I've asked over there to see what they say

**Reply by StefanOltmann** [reply]
> it does seem that putting some of the namespace declarations on the rdf node rather than the xmpmeta node would help

Yes, I would recommend that, as it aligns with the behavior of Adobe XMP SDK, ExifTool, and Exiv2.

ChatGPT also indicates that this is the expected specification, though I haven’t found a definitive source to confirm it.

I remove everything around `<rdf>` because some files contain random junk around the tag, which causes issues with the XML parsing library I use.

**Reply by Zeugma440** [reply]
Thanks @StefanOltmann (and chatGPT I guess?) for the extra hints

I have a new update ready on a local branch. It declares `xmlns:x` on the `x:xmpmeta` node and declares all other namespaces on the `rdf:RDF` node. @Numpsy Please tell me if I should push it to main.

**Reply by StefanOltmann** [reply]
Great! :)

So I'll close https://github.com/Ashampoo/xmpcore/issues/36 👍

**Reply by Zeugma440** [reply]
Changes are available on today's v6.15

**Reply by Numpsy** [reply]
> expected specification

I suppose that 'do whatever the C++ version of XmpCore does' is a reasonable starting point.

> Changes are available on today's v6.15

hanks, I;ll give it a go.

**Comment by Zeugma440** [comment]
XMP support for WAV and M4A/MP4 has been released with today's v5.21
